ob1k-core/src/main/java/com/outbrain/ob1k/client/endpoints/AsyncClientEndpoint.java

```diff
@@ -83,15 +88,58 @@ public AsyncClientEndpoint(final HttpClient httpClient, final RequestMarshallerR
     }
   }
 
-  @Override
-  public Object invoke(final TargetProvider targetProvider, final Object[] params) throws Throwable {
-    final String remoteTarget;
-    try{
+  private class InvokeAsyncAction<T> implements FutureAction<T> {
+    private final TargetProvider targetProvider;
+    private final AsyncClientEndpoint asyncClientEndpoint;
+    private final Object[] params;
+    private final DoubleDispatchStrategy doubleDispatchStrategy;
+    private volatile String firstInvocationTarget;
+
+    public InvokeAsyncAction(final TargetProvider targetProvider, final Object[] params, final DoubleDispatchStrategy doubleDispatchStrategy, final AsyncClientEndpoint asyncClientEndpoint) {
+      this.targetProvider=targetProvider;
+      this.asyncClientEndpoint = asyncClientEndpoint;
+      this.doubleDispatchStrategy = doubleDispatchStrategy;
+      this.params = params;
+    }
+
+    @Override
+    public ComposableFuture<T> execute() {
+      final String remoteTarget;
+      try{
+        remoteTarget = provideTarget();
+      } catch (final RuntimeException e) {
+        return fromError(e);
+      }
+      final DefaultAsyncClientRequestContext ctx = new DefaultAsyncClientRequestContext(remoteTarget, params, asyncClientEndpoint);
+      final long startTime = System.currentTimeMillis();
+      final ComposableFuture<T> result = asyncClientEndpoint.invokeAsync(ctx);
+      if (doubleDispatchStrategy != null) {
+        result.consume((res) -> doubleDispatchStrategy.onComplete(res, startTime));
+      }
+      return result;
+    }
+
+    private String provideTarget() {
+      String remoteTarget;
       remoteTarget = targetProvider.provideTarget();
-    } catch (final RuntimeException e) {
-      return fromError(e);
+      if (firstInvocationTarget != null) {
```

Comment:
OK, deleted my prev comments.
I think this may be correct, but it assumes too much about the calling code, and the provider behaviour, which makes it very fragile.